## Programming part of Homework 2 (Data Structures, Fall 2022)

## Name: 歐佳昀
## Student ID Number: 110590450

### Programming problem 1
Consider the problem of generating the series $S$ of numbers in Problem 3. Please use Python to implement the pseudo-code you provide for solving the problem. You should provide the iterative and recursive functions named with `S_iterative()` and `S_recursive()` respectively. Then, perform a comparison on these two versions by measuring the execution time. Besides, we also consider the number of recursive calls made in the recursive approaches. Hence, you need to
1. implement the iterative approach as a function named as `S\_iterative()`,
2. provide the recursive version for the approach named as `S\_recursive()`, and
3. compare these two function with the same input in terms of running time and write what you have observed.

We will use an in-built python library `timeit` and the module function `timeit.timeit()` for measuring the running time respective. In you observation, please indicate the ways to measure the time and this will be provided in the template.

**Note:** For the programming problems, please do some experiments by yourself to observe the running time and the number of recursive calls. You need to report what you have observed from the experiments when you submit your homework.

In [1]:
# the iterative approach
def S_iterative(n):
    now = 0
    for i in range(2,n+1,1):
        now = now + pow(2,i-2)+1
    return now

n=10
print ("S_iterative:")
print ("S_", n , ":", S_iterative(n))

S_iterative:
S_ 10 : 520


In [2]:
# the recursive version
def S_recursive(n):
    global total 
    total+=1
    
    if (n <= 2):
        return (n//2)*2
    else:
        return 3*(S_recursive(n-1)) -2*(S_recursive(n-2)) - 1

n=10
total=0
print ("S_recursive:")
print ("S_", n , ":", S_recursive(n) , "; recursive calls: " , total ,) 

S_recursive:
S_ 10 : 520 ; recursive calls:  109


In [3]:
#---------------------------------------------------------timeit
import timeit as ti
n=10

def S_iterative(n):
    now = 0
    for i in range(2,n+1,1):
        now = now + pow(2,i-2)+1
    return now

def S_recursive(n):
    global total 
    total+=1
    
    if (n <= 2):
        return (n//2)*2
    else:
        return 3*(S_recursive(n-1)) -2*(S_recursive(n-2)) - 1


#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    S_iterative(n)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    S_recursive(n)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

print("Iterative on", n, "=", S_iterative(n), "; Recursive on", n, "=", S_recursive(n))
print("iterative approach:", t_ite)
print("Recursive approach:", t_rec)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")

def iterative_test():
    S_iterative(n)
    
def recursive_test():
    S_recursive(n)
    
t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number = 10000)#timeit(函數名稱,執行次數)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number = 10000)

print("Iterative on", n, "=", S_iterative(n), "; Recursive on", n, "=", S_recursive(n))
print("iterative approach:", t_ite)
print("Recursive approach:", t_rec)


Running time comparison (Timer Method 1):
Iterative on 10 = 520 ; Recursive on 10 = 520
iterative approach: 0.04558599996380508
Recursive approach: 0.24533529998734593
Running time comparison (Timer Method 2):
Iterative on 10 = 520 ; Recursive on 10 = 520
iterative approach: 0.036760499933734536
Recursive approach: 0.23384859994985163


#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments
1. S_11 = 1033<br>
2. S_15 = 16397<br>
3. S_18 = 131088<br>
5. S_20 = 524306<br>
5. S_22 = 2097172<br>
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.024506399990059435|0.0344463000074029|0.0429594999877736|0.04790769994724542|0.05297189997509122|
|Recursion      |0.21517870004754514|1.6401275000534952|6.4269306000787765|16.279682100052014|45.10676330002025|

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.024700399953871965|0.03486410004552454|0.04292769997846335|0.04211330006364733|0.05335679999552667|
|Recursion      |0.2217032000189647|1.5236719999229535|7.107825200073421|18.610733599984087|45.6011029999936|

*******************************************************************************************************

##### Observations


根據觀察，我有下列幾點發現:

1. Iteration 和 Recursion 的時間比較上，iteration 運算所需的時間皆比recusion 來的要短很多，這跟函數的運行次數有關，透過以上代碼中，我們可以發現，在S_10 的時候，Iteration 裡迭代次數為9次(for loop 內的執行次數,n= 0 不須進入迴圈)，而使用 Recursion 的方式卻需要高達109次才可遞迴出答案，而n的數字越大，需要遞迴的次數越多，而Iteration仍然只需n-1 次即可算出答案(ex. 當S_18 時，Recursion需遞迴約5167次，Iteration只需迭代17次 )，故Recursion 所需時間才會越來越大

Iteration 迭代次數 I(n) : if n>1 :In(n) =  n-1 <br>
Recursion 遞迴次數 R(n) : if n <= 2 : R(n) = 1, if n >2 :R(n) = R(n-1) + R(n-2) + 1 <br>
(據觀察得出)


2. method 1 和 method 2 的時間花費不同，是因為 method 1 是使用for loop 的方式，並將結束時間扣除開始時間得出(但timeti.default_timer()為整個系統時間，其他程序出現可能會干擾測量)，而method 2 的 timeit 則是將同樣的程式執行 10000 次並提供從給定的程式碼片段集中獲取的最短時間返還

### Programming problem 2
A ***permutation*** $P$ of size $n$ is a way of arranging the numbers between 1 and $n$, where each number must be used and can only be used once. For example, 5 4 1 2 3 is a permutation of 1 2 3 4 5 where $P(5)=1$, $P(4)=2$, $P(1)=3$, $P(2)=4$, and $P(3)=5$. An ***inversion*** in a permutation $P$ is a pair of numbers $i \mbox{ and} j \in\{1, \ldots, n\}$ with $i<j$ such that the larger number appears to the left of the smaller one in the permutation, that is $P(j)<P(i)$. In the example, the pair of (3, 5) is an inversion in $P$. The ***inversion number*** of a permutation is the total number of inversions. Please write the Python fuctions mentioned in Problem 4 and compare the performance in term of running time. The things to be provided are as below.
1. Implement the iterative approach using Python and name the function as `inversion_number_iterative()`.
2. Please implement the linear recursive version you provide in Problem 4(c) with the function name of `inversion\_number\_recursive()`.
3. Write the binary recursive version you provide in Problem 4(d) with the function named as `inversion\_number\_tworecurs()`.
4. Compare these functions with the same input in terms of running time and write what you have observed.

In [1]:
# the iterative approach
def inversion_number_iterative(list_tmp):
    n = 0
    for i in range(len(list_tmp)-1):
        for j in range(i+1,len(list_tmp),1):
            if(list_tmp[i] > list_tmp[j]):
                n = n +  1

    return n


#perm=[4, 6, 2, 5, 1, 3]
#perm = [12, 11, 13, 5, 6, 7]
perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_iterative(perm))

The inversion number of  [35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82] is 19


In [5]:
# the linear recursive version
def inversion_number_recursive(list_tmp):
    return  linear_recursive_start(list_tmp,0)

def linear_recursive_start(list_tmp,n):
    count = 0
    if n+1 == len(list_tmp):
        return 0
    else:
        for i in range(n+1,len(list_tmp),1):
            if list_tmp[n] > list_tmp[i]:
                count+=1
        # list_tmp = list(list_tmp)
        # list_tmp.pop(0)
        return count + linear_recursive_start(list_tmp,n+1) # n = n+1 => iterative i+1




#perm=[4, 6, 2, 5, 1, 3]
#perm = [12, 11, 13, 5, 6, 7]
perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_recursive(perm))

The inversion number of  [35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82] is 19


In [2]:
# the binary recursive version


def inversion_number_tworecurs(arr):

    return merge_sort(list(tuple(arr)), 0, len(arr)-1)

def merge_sort(arr_tmp, left, right):

    count = 0

    if left < right:
        mid = (left + right)//2

        count += merge_sort(arr_tmp, left, mid) + merge_sort(arr_tmp, mid+1, right)

        tmp_arr = []
        tmp_count = 0
        tmp_start_l = left
        tmp_start_r = mid+1

        while tmp_start_l <= mid and tmp_start_r <= right:

            if (arr_tmp[tmp_start_l] <= arr_tmp[tmp_start_r]):

                tmp_arr.append(arr_tmp[tmp_start_l])
                tmp_start_l += 1
                n = tmp_start_r

            else:
                tmp_arr.append(arr_tmp[tmp_start_r])
                tmp_start_r += 1
                tmp_count += mid-tmp_start_l + 1
                n = tmp_start_l

        m = right if n == tmp_start_r else mid
        n = tmp_start_r if m == right else tmp_start_l

        tmp_arr.extend(arr_tmp[n:m+1])
        arr_tmp[left:right+1] = tmp_arr[:len(tmp_arr)]

        return tmp_count + count

    return count


#perm=[4, 6, 2, 5, 1, 3]
#perm = [12, 11, 13, 5, 6, 7]
perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
print('The inversion number of ', perm, 'is', inversion_number_tworecurs(perm))

The inversion number of  [35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82] is 19


In [4]:

import timeit as ti

################ 1

def inversion_number_iterative(list_tmp):
    n = 0
    for i in range(len(list_tmp)-1):
        for j in range(i,len(list_tmp),1):
            if(list_tmp[i] > list_tmp[j]):
                n = n +  1

    return n


################ 2

def inversion_number_recursive(list_tmp):
    return  linear_recursive_start(list_tmp,0)

def linear_recursive_start(list_tmp,n):
    count = 0
    if n+1 == len(list_tmp):
        return 0
    else:
        for i in range(n,len(list_tmp),1):
            if list_tmp[n] > list_tmp[i]:
                count+=1
        # list_tmp = list(list_tmp)
        # list_tmp.pop(0)
        return count + linear_recursive_start(list_tmp,n+1) # n = n+1 => iterative i+1



################ 3

def inversion_number_tworecurs(arr):

    return merge_sort(list(tuple(arr)), 0, len(arr)-1)


def merge_sort(arr_tmp, left, right):

    count = 0

    if left < right:
        mid = (left + right)//2

        count += merge_sort(arr_tmp, left, mid) + merge_sort(arr_tmp, mid+1, right)

        tmp_arr = []
        tmp_count = 0
        tmp_start_l = left
        tmp_start_r = mid+1

        while tmp_start_l <= mid and tmp_start_r <= right:

            if (arr_tmp[tmp_start_l] <= arr_tmp[tmp_start_r]):

                tmp_arr.append(arr_tmp[tmp_start_l])
                tmp_start_l += 1
                n = tmp_start_r

            else:
                tmp_arr.append(arr_tmp[tmp_start_r])
                tmp_start_r += 1
                tmp_count += mid-tmp_start_l + 1
                n = tmp_start_l

        m = right if n == tmp_start_r else mid
        n = tmp_start_r if m == right else tmp_start_l
        
        for i in range(n,m+1,1):
            tmp_arr.append(arr_tmp[i])
            
        arr_tmp[left:right+1] = tmp_arr[:len(tmp_arr)]

        return tmp_count + count

    return count


#perm = [12, 11, 13, 5, 6, 7]
perm=[35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")

tstart = ti.default_timer() #timer start 計時器開啟

for i in range(10000):
    inversion_number_iterative(perm)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    inversion_number_recursive(perm)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    
    # print(inversion_number_tworecurs(perm))
    inversion_number_tworecurs(perm)
tend = ti.default_timer() #timer end計時器關閉
t_twr=tend - tstart#開始與結束的差值


print("Permutation:", perm)
print('The inversion number of is', inversion_number_iterative(perm),'(iterat)')
print('The inversion number of is', inversion_number_recursive(perm),'(linear)')
print('The inversion number of is', inversion_number_tworecurs(perm),'(binary)')

print("Iterative approach:", t_ite)
print("Recursive approach:", t_rec)
print("Binrecurs approach:", t_twr)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")

def iterative_test():
    inversion_number_iterative(perm)

def recursive_test():
    inversion_number_recursive(perm)


def binrecurs_test():
    inversion_number_tworecurs(perm)
    
t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number = 10000)#timeit(函數名稱,執行次數)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number = 10000)
t_twr = ti.timeit("binrecurs_test()", setup="from __main__ import binrecurs_test", number = 10000)

print("Iterative approach:", t_ite)
print("Recursive approach:", t_rec)
print("Binrecurs approach:", t_twr)

Running time comparison (Timer Method 1):
Permutation: [35, 11, 26, 13, 64, 21, 44, 6, 100, 57, 77, 82]
The inversion number of is 19 (iterat)
The inversion number of is 19 (linear)
The inversion number of is 19 (binary)
Iterative approach: 0.11682869994547218
Recursive approach: 0.1288813001010567
Binrecurs approach: 0.2828361999709159
Running time comparison (Timer Method 2):
Iterative approach: 0.10179310000967234
Recursive approach: 0.12961970001924783
Binrecurs approach: 0.271156599977985


#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments
1.perm = [10,9,8,7,6,5,4,3,2,1] = 45<br>
2.perm = [20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1] = 190<br>
3.perm = [30,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1] = 435<br>
4.perm = [40,39,38,37,36,35,34,33,32,31,30,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1] = 780<br>
5.perm = [45,44,43,42,41,40,39,38,37,36,35,34,33,32,31,30,29,28,27,26,25,24,23,22,21,20,19,18,17,16,15,14,13,12,11,10,9,8,7,6,5,4,3,2,1] = 990<br>
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.03434379992540926|0.10556040005758405|0.22869659995194525|0.39110380003694445|0.48435239994432777|
|Recursion      |0.04427650000434369|0.1320225999224931|0.27070210000965744|0.4512387999566272|0.5489717000164092|
|Binrecurs      |0.07759090000763535|0.17628530005458742|0.296524400007911|0.39588160009589046|0.4605821999721229|

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      |0.034476300003007054|0.10695899999700487|0.2292741000419482|0.4081772000063211|0.5129084999207407|
|Recursion      |0.04442839999683201|0.13133539992850274|0.2652766000246629|0.45712699997238815|0.5474417000077665|
|Binrecurs      |0.07843059999868274|0.18120919994544238|0.2937742000212893|0.3892200000118464|0.4666822999715805|

*******************************************************************************************************

##### Observations


透過上述對程式碼的觀察，我有以下發現:

1.分別設置將會在同樣長度n 的情況下會產生最大逆序數之降序陣列後，我們可以發現，一開始使用Iteration的方式，效率會比Recursion 和 Binrecurs 快，而隨著陣列長度不斷增加，Iteration的效率會越來越慢，Recursion和Iteration相差無幾，最後Binrecurs的效率會約略在n達到45時同時在method1、methodf2 中成為三種算法中最快的算法

2.透過對時間複雜度與執行步驟的分析，我們可以得出Iteration 和 Recursion 的執行步驟數差不多，並經計算後可得出 Binrecurs 的時間複雜度為 O(nlogn)，而Iteration 則是 O(n^2)，所以雖然Binrecurs 過程繁瑣且起初效率並不高，但當n大到一定程度時，即可使得效率優於Iteration